src/lib/select/select.ts

```diff
@@ -456,68 +531,94 @@ export class MdSelect implements AfterContentInit, ControlValueAccessor, OnDestr
   /** Sets up a key manager to listen to keyboard events on the overlay panel. */
   private _initKeyManager() {
     this._keyManager = new FocusKeyManager(this.options);
-    this._tabSubscription = this._keyManager.tabOut.subscribe(() => {
-      this.close();
-    });
+    this._tabSubscription = this._keyManager.tabOut.subscribe(() => this.close());
   }
 
   /** Drops current option subscriptions and IDs and resets from scratch. */
   private _resetOptions(): void {
     this._dropSubscriptions();
     this._listenToOptions();
     this._setOptionIds();
+    this._setOptionMultiple();
   }
 
-  /** Listens to selection events on each option. */
+  /** Listens to user-generated selection events on each option. */
   private _listenToOptions(): void {
-    this.options.forEach((option: MdOption) => {
-      const sub = option.onSelect.subscribe((event: MdOptionSelectEvent) => {
-        if (event.isUserInput && this._selected !== option) {
-          this._emitChangeEvent(option);
+    this._optionSubscription = this.optionSelectionChanges
+      .filter(event => event.isUserInput)
+      .subscribe(event => {
+        this._onSelect(event.source);
+        this._setValueWidth();
+
+        if (!this.multiple) {
+          this.close();
         }
-        this._onSelect(option);
       });
-      this._subscriptions.push(sub);
-    });
+  }
+
+  /** Invoked when an option is clicked. */
+  private _onSelect(option: MdOption): void {
+    const wasSelected = this._selectionModel.isSelected(option);
+
+    if (this.multiple) {
+      this._selectionModel.toggle(option);
+    } else {
+      this._clearSelection(option);
+      this._selectionModel.select(option);
+    }
+
+    this._sortValues();
+
+    if (wasSelected !== this._selectionModel.isSelected(option)) {
+      this._propagateChanges();
+    }
+  }
+
+  /**
+   * Sorts the model values, ensuring that they keep the same
+   * order that they have in the panel.
+   */
+  private _sortValues(): void {
+    const options = this.options.toArray();
+    this._selectionModel.sort((a, b) => options.indexOf(a) - options.indexOf(b));
```

Comment:
That may complicate the selection model even more than adding the `emitChangeEvent` to `select`, `deselect` and `clear`. In general, I doubt that we'd run into performance issues with this in particular, because:

1. Even if we had hundreds of items, sorting through them would still take less than a millisecond.
2. The sorting only happens on click, it's not something that we'd be triggering on each change detection cycle.